<a href="https://colab.research.google.com/github/Data-Analytics-with-Python/individual-assignment-iii-javedt3/blob/main/Assignment_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

IMPORTANT: Before you start, enter your name and student number below.

**Full Name**:

**Student Number**:

# Predictive Analytics for Nata Supermarket

Welcome to Part III of our case assignment. In this part, we will continue working with the same dataset of **Nata Supermarket**.  Our focus here will be on performing predictive modeling tasks.

Throughout this assignment, please ensure that your results are reproducible by setting the **random_state to 20**



# Loading and preparing the data

To begin with, load the data as a `pandas` data frame. Recall that you there are missing values in the data. **Make sure to address** the following issues from part I of the assignment before starting your analysis:

* Remove the missing values
* Remove any column of constant values
* Convert the column `Dt_Customer` to number of days the customer has been with the company.

In [89]:
import pandas as pd

df = pd.read_csv("/W33836-XLS-ENG (2).csv")

df.dropna(inplace=True)

df

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0  2012-04-09       58       635  ...                  7   
1            1  2014-08-03       38        11  ...                  5   
2            0  21-08-2013       26       426  ...                  4   
3            0  2014-10-02       26        11  ...                  6   
4            0  19-01-2014       94       173  ...                  5   
...        ...         ...      ...       ...  ...                ...   
2235         1  13-06-2013       46       709  ...                  5   
2236         1  2014-10-06       56       406  ...                  7   
2237         0  25-01-2014       91       908  ...                  6   
2238         1  24-01-2014        8       428  ...                  3   
2239         1  15-10-2012       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
1            0              3         11         0  
2            0              3         11         0  
3            0              3         11         0  
4            0              3         11         0  
...        ...            ...        ...       ...  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[2216 rows x 29 columns]

In [90]:
variance = df.select_dtypes(include=['number']).var()
constant_columns = variance[variance == 0].index.tolist()
df.drop(columns=constant_columns, inplace=True)
df

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumCatalogPurchases  \
0            0  2012-04-09       58       635  ...                   10   
1            1  2014-08-03       38        11  ...                    1   
2            0  21-08-2013       26       426  ...                    2   
3            0  2014-10-02       26        11  ...                    0   
4            0  19-01-2014       94       173  ...                    3   
...        ...         ...      ...       ...  ...                  ...   
2235         1  13-06-2013       46       709  ...                    3   
2236         1  2014-10-06       56       406  ...                    2   
2237         0  25-01-2014       91       908  ...                    3   
2238         1  24-01-2014        8       428  ...                    5   
2239         1  15-10-2012       40        84  ...                    1   

      NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  \
0                     4                  7             0             0   
1                     2                  5             0             0   
2                    10                  4             0             0   
3                     4                  6             0             0   
4                     6                  5             0             0   
...                 ...                ...           ...           ...   
2235                  4                  5             0             0   
2236                  5                  7             0             0   
2237                 13                  6             0             1   
2238                 10                  3             0             0   
2239                  4                  7             0             0   

      AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  Response  
0                0             0             0         0         1  
1                0             0             0         0         0  
2                0             0             0         0         0  
3                0             0             0         0         0  
4                0             0             0         0         0  
...            ...           ...           ...       ...       ...  
2235             0             0             0         0         0  
2236             0             1             0         0         0  
2237             0             0             0         0         0  
2238             0             0             0         0         0  
2239             0             0             0         0         1  

[2216 rows x 27 columns]

In [91]:
from datetime import datetime

df["Dt_Customer"] = pd.to_datetime(df["Dt_Customer"], format='mixed', dayfirst=True)
today = datetime.today()
df['Customer_Days'] = (today - df["Dt_Customer"]).dt.days

display(df['Customer_Days'])

0       4980
1       4134
2       4481
3       4074
4       4330
        ... 
2235    4550
2236    4070
2237    4324
2238    4325
2239    4791
Name: Customer_Days, Length: 2216, dtype: int64

# Question 1 (45 points)

In this question, we will predict a customer's spending amount on each product category over a two year period. Let us assume that when we try to predict a customer's spending on a product category (such as wines), their spending on other products is not observable.

In this question and Question 2, we will focus on **wines**.

(i). **Split the data** into two data frames, X (**features**) and y (**target**).

Then, further **split the data** into **training** and **testing** sets.

In [92]:

y = df['MntWines']
X = df.drop(columns=['MntWines', 'ID', 'Dt_Customer'])

display(X.head())
display(y.head())

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  Recency  \
0        1957  Graduation         Single  58138.0        0         0       58   
1        1954  Graduation         Single  46344.0        1         1       38   
2        1965  Graduation       Together  71613.0        0         0       26   
3        1984  Graduation       Together  26646.0        1         0       26   
4        1981         PhD        Married  58293.0        1         0       94   

   MntFruits  MntMeatProducts  MntFishProducts  ...  NumStorePurchases  \
0         88              546              172  ...                  4   
1          1                6                2  ...                  2   
2         49              127              111  ...                 10   
3          4               20               10  ...                  4   
4         43              118               46  ...                  6   

   NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  \
0                  7             0             0             0             0   
1                  5             0             0             0             0   
2                  4             0             0             0             0   
3                  6             0             0             0             0   
4                  5             0             0             0             0   

   AcceptedCmp2  Complain  Response  Customer_Days  
0             0         0         1           4980  
1             0         0         0           4134  
2             0         0         0           4481  
3             0         0         0           4074  
4             0         0         0           4330  

[5 rows x 25 columns]

0    635
1     11
2    426
3     11
4    173
Name: MntWines, dtype: int64

In [93]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=20)

(ii). **Build a machine learning pipeline** that combines the following steps to predict spending amount on wines:

* Performing one-hot encoding for the categorical features;
* A random forest model for regression.

In the random forest model, **specify** the following hyperparameters:
* Number of trees;
* Maximum depth of any tree
* Minimum number of data points required to split a node;
* Minimum number of data points in any leaf node

In addition, **fit your model** to the training data.

In [94]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestRegressor
from sklearn.pipeline import Pipeline

categorical_features = ['Education', 'Marital_Status']
numerical_features = [col for col in X_train.columns if col not in categorical_features]

ct = ColumnTransformer(transformers=[("cat", OneHotEncoder(drop="first", sparse_output=False, handle_unknown='ignore'), categorical_features)
    ],remainder="passthrough")

rf = RandomForestRegressor(
    n_estimators=100,
    max_depth=10,
    min_samples_split=5,
    min_samples_leaf=2,
    random_state=20
)

pipe = Pipeline(
    steps=[
        ("preprocess", ct),
        ("model", rf)
    ]
)


pipe.fit(X_train, y_train)

pipe

/usr/local/lib/python3.12/dist-packages/sklearn/compose/_column_transformer.py:1667: FutureWarning:


The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).




Pipeline(steps=[('preprocess',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('cat',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore',
                                                                sparse_output=False),
                                                  ['Education',
                                                   'Marital_Status'])])),
                ('model',
                 RandomForestRegressor(max_depth=10, min_samples_leaf=2,
                                       min_samples_split=5, random_state=20))])

(iii). Use the model to **predict** the spending amount on wines by a customer with the following features.

| Feature |  Value  |
|---------|---------|
| Age     | 48      |
|Education|Graduation|
|Marital_Status| Married|
|Income|80,000|
|Kidhome|1|
|Teenhome|1|
|Dt-Customer|2016-10-10|
|Recency|43|     
|NumDealsPurchases|2|
|   NumWebPurchases|1|
|NumCatalogPurchases|0|
|NumStorePurchases|15|
|NumWebVisitsMonth|5|
|AcceptedCmp1,2,3,4,5| 0 |
|Complain|0|




In [95]:
from datetime import datetime

new_customer_features = {
    'Education': 'Graduation',
    'Marital_Status': 'Married',
    'Income': 80000.0,
    'Kidhome': 1,
    'Teenhome': 1,
    'Recency': 43,
    'NumDealsPurchases': 2,
    'NumWebPurchases': 1,
    'NumCatalogPurchases': 0,
    'NumStorePurchases': 15,
    'NumWebVisitsMonth': 5,
    'AcceptedCmp1': 0,
    'AcceptedCmp2': 0,
    'AcceptedCmp3': 0,
    'AcceptedCmp4': 0,
    'AcceptedCmp5': 0,
    'Complain': 0,
    'Response': 0,
    'MntFruits': 0,
    'MntMeatProducts': 0,
    'MntFishProducts': 0,
    'MntSweetProducts': 0,
    'MntGoldProds': 0
}

new_customer_features['Year_Birth'] = 2016 - 48

dt_customer_new = pd.to_datetime('2016-10-10')
today = datetime.today()
new_customer_features['Customer_Days'] = (today - dt_customer_new).days

new_customer_df = pd.DataFrame([new_customer_features])

print("New customer DataFrame created successfully:")
display(new_customer_df)

New customer DataFrame created successfully:


Education Marital_Status   Income  Kidhome  Teenhome  Recency  \
0  Graduation        Married  80000.0        1         1       43   

   NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
0                  2                1                    0                 15   

   ...  AcceptedCmp5  Complain  Response  MntFruits  MntMeatProducts  \
0  ...             0         0         0          0                0   

   MntFishProducts  MntSweetProducts  MntGoldProds  Year_Birth  Customer_Days  
0                0                 0             0        1968           3335  

[1 rows x 25 columns]

In [96]:
predicted_mntwines = pipe.predict(new_customer_df)

predicted_mntwines


array([381.17025866])

(iv). **Consider** **two** measures to evaluate the model's performance on the test dataset.

Based on you computational results, how would you describe the model's performance?

In [97]:
from sklearn.metrics import mean_absolute_error, r2_score

y_pred = pipe.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R-squared (R2): {r2:.2f}")

Mean Absolute Error (MAE): 89.67
R-squared (R2): 0.78


Using Mean Absolute Error (MAE) and R² to evaluate the model, the results indicate strong predictive performance on the test data. The model achieves an MAE of 89.67, which means that the predicted wine-spending amounts differ from the actual values by about 90 dollars on average, indicating reasonably accurate individual predictions. The R² value of 0.78 shows that the model explains 78% of the variance in wine spending, suggesting that the model captures the main patterns in the data well.

Overall, the model performs well both in terms of prediction accuracy and goodness-of-fit, although some prediction error remains—particularly for extreme values. Nevertheless, the results indicate that the model is reliable and effective for predicting customers’ wine spending.

(v). **Perform** a 6-fold cross validation with a performance score of your choice.

**Note**: You may need to research on how to specify the performance score for regression models.

In [98]:
from sklearn.model_selection import KFold, cross_val_score

kf = KFold(n_splits=6, shuffle=True, random_state=20)

scoring_metric = 'neg_mean_absolute_error'

scores = cross_val_score(pipe, X_train, y_train, cv=kf, scoring=scoring_metric)

print(f"Individual cross-validation scores ({scoring_metric}): {scores}")

print(f"Mean cross-validation score ({scoring_metric}): {scores.mean():.2f}")
print(f"Standard deviation of cross-validation scores ({scoring_metric}): {scores.std():.2f}")

print(f"Mean Absolute Error (CV): {-scores.mean():.2f}")

/usr/local/lib/python3.12/dist-packages/sklearn/preprocessing/_encoders.py:246: UserWarning:

Found unknown categories in columns [1] during transform. These unknown categories will be encoded as all zeros



Individual cross-validation scores (neg_mean_absolute_error): [-96.32562554 -96.09694881 -83.18071117 -80.74221473 -78.60150318
 -79.92909667]
Mean cross-validation score (neg_mean_absolute_error): -85.81
Standard deviation of cross-validation scores (neg_mean_absolute_error): 7.48
Mean Absolute Error (CV): 85.81


(vi). **Perform** hyperparameter tuning using `GridSearchCV` for the following hyperparameters:

* Number of trees: 50, 100
* Maximum depth of any tree: 5, 10, 15
* Minimum number of data points required to split a node: 3, 6
* Minimum number of data points in any leaf node: 2,4,8


Based on your computational result, **show**:
* the best hyperparameter comination
* the corresponding performance score

In addition, **retrieve** the best model (the one corresponding to the best performance score).

In [99]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'model__n_estimators': [50, 100],
    'model__max_depth': [5, 10, 15],
    'model__min_samples_split': [3, 6],
    'model__min_samples_leaf': [2, 4, 8]
}

grid_search = GridSearchCV(
    estimator=pipe,
    param_grid = param_grid,
    cv = kf,
    scoring='neg_mean_absolute_error',
    n_jobs=-1,
    verbose=1
)

grid_search.fit(X_train, y_train)



Fitting 6 folds for each of 36 candidates, totalling 216 fits


/usr/local/lib/python3.12/dist-packages/sklearn/compose/_column_transformer.py:1667: FutureWarning:


The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).




GridSearchCV(cv=KFold(n_splits=6, random_state=20, shuffle=True),
             estimator=Pipeline(steps=[('preprocess',
                                        ColumnTransformer(remainder='passthrough',
                                                          transformers=[('cat',
                                                                         OneHotEncoder(drop='first',
                                                                                       handle_unknown='ignore',
                                                                                       sparse_output=False),
                                                                         ['Education',
                                                                          'Marital_Status'])])),
                                       ('model',
                                        RandomForestRegressor(max_depth=10,
                                                              min_samples_leaf=2,
                                                              min_samples_split=5,
                                                              random_state=20))]),
             n_jobs=-1,
             param_grid={'model__max_depth': [5, 10, 15],
                         'model__min_samples_leaf': [2, 4, 8],
                         'model__min_samples_split': [3, 6],
                         'model__n_estimators': [50, 100]},
             scoring='neg_mean_absolute_error', verbose=1)

In [100]:
grid_search.best_params_


{'model__max_depth': 15,
 'model__min_samples_leaf': 2,
 'model__min_samples_split': 3,
 'model__n_estimators': 100}

In [101]:
print(f"Corresponding best performance score (Mean Absolute Error): {-grid_search.best_score_:.2f}")


Corresponding best performance score (Mean Absolute Error): 84.07


# Question 2 (24 points)

In this question, we will compare the performance of the best model found through `GridSearchCV` in Question 1 with the performance of the linear regression model.

(i). **Construct** a linear regression model using all relevant features and fit it to the training data.

Further, **evaluate** the model's performance on the test data and compare it with the best random forest model found in Question 1, with respect to the two performance considered in Question 1.

**Note**: You may use the same training and testing datasets as in Question 1.

In [102]:
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_absolute_error, r2_score

linear_regression_pipeline = Pipeline(steps=[
    ('preprocess', ct),
    ('regressor', LinearRegression())
])

linear_regression_pipeline.fit(X_train, y_train)

y_pred_lr = linear_regression_pipeline.predict(X_test)


mae_lr = mean_absolute_error(y_test, y_pred_lr)
r2_lr = r2_score(y_test, y_pred_lr)

print(f"Linear Regression MAE: {mae_lr:.2f}")
print(f"Linear Regression R2: {r2_lr:.2f}")

y_pred_rf = grid_search.best_estimator_.predict(X_test)

mae_rf = mean_absolute_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print(f"Random Forest MAE: {mae_rf:.2f}")
print(f"Random Forest R2: {r2_rf:.2f}")


Linear Regression MAE: 119.71
Linear Regression R2: 0.67
Random Forest MAE: 87.93
Random Forest R2: 0.78


(ii). Let's further compare the distribution of prediction errors by the two models in the following steps.

**Step 1**. For both the linear regression and the (best) random forest model, compute the absolute residual residual for each record in the test dataset.
  * Note that the absolute residual is distance between the predicted value and actual value, i.e., $|y_{pred}-y_{test}|$.

So we end up with two sets of absolute residuals (one by the linear regression model and the other by the random forest model).

**Step 2**. For each pair of absolute residuals for the same test data point, we can define a point in a scatterplot. Genereate such a scatterplot using `plotly.express` (LR residuals vs. RF residuals).  

**Step 3**. Add a 45 degree reference line to the plot. This can be done using the following codes. (You may need to change `min_val` and `max_val` for better visualization).

```
min_val = 0
max_val = 10
fig.add_shape(
    type="line",
    x0=min_val, y0=min_val,
    x1=max_val, y1=max_val,
    line=dict(color="red", width=2, dash="dash"),
)
```

**Implement the above steps**.

Note that the above steps essentially creates a [Q-Q plot](https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot). How would you interpret the plot (for comparing the two predictive models).

In [103]:
import numpy as np
import pandas as pd

abs_resid_lr = np.abs(y_pred_lr - y_test)
abs_resid_rf = np.abs(y_pred_rf - y_test)

resid_df = pd.DataFrame({
    "LR_abs_resid": abs_resid_lr,
    "RF_abs_resid": abs_resid_rf
})

display(resid_df.head())


LR_abs_resid  RF_abs_resid
1475     41.780191      3.122619
1108     80.807300     89.346595
493     376.470915    157.319012
1370     16.384684     22.201226
2103      3.092634     42.957373

In [104]:
import plotly.express as px

fig = px.scatter(
    resid_df,
    x="LR_abs_resid",
    y="RF_abs_resid",
    labels={
        "LR_abs_resid": "Linear Regression |Residual|",
        "RF_abs_resid": "Random Forest |Residual|"
    },
    title="Absolute Residuals: Linear Regression vs Random Forest"
)

fig.show()


In [105]:
min_val = 0
max_val = max(resid_df["LR_abs_resid"].max(),
              resid_df["RF_abs_resid"].max()) * 1.05   # slight padding

fig.add_shape(
    type="line",
    x0=min_val, y0=min_val,
    x1=max_val, y1=max_val,
    line=dict(color="red", width=2, dash="dash"),
)

fig.update_xaxes(range=[min_val, max_val])
fig.update_yaxes(range=[min_val, max_val])

fig.show()


This plot compares the absolute prediction errors of the linear regression model (x-axis) with those of the random forest model (y-axis) for each customer in the test set. The dashed red 45-degree line represents cases where both models make the same size error on a given observation. Points below the line indicate that the random forest is more accurate than the linear regression for that customer. Points above the line indicate that linear regression performs better for that customer. Points on the line represent equal performance by both models. In this plot, the majority of points lie below the reference line, showing that the random forest generally has smaller absolute residuals than linear regression for most customers. This confirms that the random forest model provides more accurate predictions overall. Additionally, for observations where linear regression makes large errors (far to the right on the x-axis), random forest often shows much smaller corresponding errors (lower y-values). This indicates that random forest is not only better on average, but also more accurate for extreme cases, where linear regression tends to perform poorly. A small number of points appear above the line, meaning there are cases where linear regression performs better, but these are relatively limited and tend to be small-error cases.

# Question 3. (24 points)

In this question, we will consider a classification problem on customers' spendings on meat products.


(i). For the column that represents customers' spendings on meat products, **calculate** the 33.33% and 66.67% percentiles. (**Hint**: You may use the function `df['MntMeatProducts'].quantile([1/3,2/3])`.)

Based on the two percentiles, **label** each row in the dataset
* If a customer's spending is below the 33.33% percentile, label their spending as "low";

* If a customer's spending is above the 66.67% percentile, label their spending as "high";


* If a customer's spending is between the two percentiles, label their spending as "medium".

In [106]:
percentiles = df['MntMeatProducts'].quantile([1/3, 2/3])

p33 = percentiles.loc[1/3]
p67 = percentiles.loc[2/3]

print("33.33% percentile:", p33)
print("66.67% percentile:", p67)


33.33% percentile: 24.0
66.67% percentile: 146.33333333333258


In [107]:
def label_meat_spending(x):
    if x < p33:
        return "low"
    elif x > p67:
        return "high"
    else:
        return "medium"

df['MeatSpendingLabel'] = df['MntMeatProducts'].apply(label_meat_spending)

df['MeatSpendingLabel'].value_counts()


MeatSpendingLabel
medium    746
high      739
low       731
Name: count, dtype: int64

In [139]:
df[['MntMeatProducts', 'MeatSpendingLabel']].head(10)


MntMeatProducts MeatSpendingLabel
0              546              high
1                6               low
2              127            medium
3               20               low
4              118            medium
5               98            medium
6              164              high
7               56            medium
8               24            medium
9                6               low

\(ii). **Build a K-nearest-neighbors model** to predict the label of a customer's spending on meat products. The model should be part of a machine learning pipeline that preprocesses the data.





In [164]:
from sklearn.neighbors import KNeighborsClassifier

y_meat = df['MeatSpendingLabel']
X_meat = df.drop(columns=['MntMeatProducts', 'ID', 'Dt_Customer', 'MeatSpendingLabel'])

X_meat_train, X_meat_test, y_meat_train, y_meat_test = train_test_split(X_meat, y_meat, test_size=0.2, random_state=20, stratify=y_meat)

categorical_features_meat = ['Education', 'Marital_Status']
numerical_features_meat = [col for col in X_meat_train.columns if col not in categorical_features_meat]

ct_meat = ColumnTransformer(
    transformers=[
        ("cat", OneHotEncoder(sparse_output=False, drop="first", handle_unknown='ignore'), categorical_features_meat),
    ],
    remainder="passthrough")

knn = KNeighborsClassifier()

pipeline_knn = Pipeline(steps=[
    ('preprocess', ct_meat),
    ('classifier', knn)
])

pipeline_knn.fit(X_meat_train, y_meat_train)



/usr/local/lib/python3.12/dist-packages/sklearn/compose/_column_transformer.py:1667: FutureWarning:


The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).




Pipeline(steps=[('preprocess',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('cat',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore',
                                                                sparse_output=False),
                                                  ['Education',
                                                   'Marital_Status'])])),
                ('classifier', KNeighborsClassifier())])

(iii). Evaluate the performace of your model in part (ii) on the test data by **generating the classification report**.

Further, **interpret** each number in the classification report based on the current context.

In [165]:
from sklearn.metrics import classification_report

y_meat_pred = pipeline_knn.predict(X_meat_test)

report = classification_report(y_meat_test, y_meat_pred)

print(report)


              precision    recall  f1-score   support

        high       0.79      0.84      0.81       148
         low       0.72      0.80      0.76       146
      medium       0.58      0.49      0.53       150

    accuracy                           0.71       444
   macro avg       0.70      0.71      0.70       444
weighted avg       0.70      0.71      0.70       444



This classification report evaluates the performance of the K-nearest neighbors (KNN) model in predicting customers’ meat-spending categories (low, medium, high) using precision, recall, F1-score, and overall accuracy.

High spenders

Precision = 0.79
When the model predicts that a customer is a high spender, it is correct 79% of the time, indicating few low or medium customers are incorrectly labeled as high.

Recall = 0.84
The model successfully identifies 84% of all true high spenders, meaning most high-value customers are captured.

F1-score = 0.81
This is the strongest performing class, reflecting an effective balance between precision and recall.

Low spenders

Precision = 0.72
Among customers predicted to be low spenders, 72% are actually low spenders.

Recall = 0.80
The model correctly labels 80% of all true low spenders.

F1-score = 0.76
Performance is strong and only slightly below the high-spending group.

Medium spenders

Precision = 0.58
Only 58% of customers predicted to be medium are truly medium.

Recall = 0.49
Fewer than half of all medium spenders are correctly identified.

F1-score = 0.53
This is the weakest class, indicating difficulty distinguishing medium spenders from adjacent categories.

Overall performance

Accuracy = 0.71
The model correctly classifies 71% of all customers.

Macro average F1 = 0.70
Average performance treating all classes equally.

Weighted average F1 = 0.70
Overall performance accounting for class sizes.


The KNN model performs well for customers with extreme spending behaviors (high and low), but struggles to accurately classify customers with medium spending levels. This is expected for a distance-based model, as medium values lie near decision boundaries in feature space. Overall, the model is effective for identifying high-value and budget-conscious customers but less reliable for customers in the mid-range.

# Describe how you used Gen. AI. in this assignment (2 points)

I have used Google Colab AI to explain errors in my code and fix them if necessary, I have also asked AI how to help create some parts of the codes for the K-N model for question 3. Overall I used AI as an aid to help me out when I absolutely needed too and could not trouble shoot the code using past assignments and lectures.

**Note**: The remaining 5 points will be assigned to readability of the work.